### ISSUE TITLE
[P1 | ROI:200x] Query-Based Scene Understanding

### ISSUE THREAD

**Issue by MikkoParkkola** [MikkoParkkola/axterminator#4]
Source: https://deepmind.google/blog/d4rt-teaching-ai-to-see-the-world-in-four-dimensions/

## Context

Current axterminator screen understanding requires a full analysis pass for every query — element location, OCR, change detection each trigger independent processing. D4RT's query-based architecture demonstrates that a single scene encoding can serve many parallel queries at 300x speedup. Applied to screen understanding, this means: encode the screen once, then answer unlimited questions about it in parallel. This is the difference between 500ms per action (current) and 50ms per action (query-based).

## Innovation
D4RT uses a query-based approach: "Where is pixel X in 3D space at time T from camera C?" Queries are independent = parallel processing. Single unified model replaces patchwork of depth/motion/pose estimators. 300x faster than SOTA.

## 100x Supercharge: Unified Screen Query System

```python
class ScreenQueryEngine:
    """Query-based screen understanding inspired by D4RT"""
    
    def __init__(self):
        self.encoder = UnifiedScreenEncoder()
        self.decoder = LightweightQueryDecoder()
    
    def encode_screen(self, screenshot: Image) -> SceneRepresentation:
        # Single encoding pass for all queries
        return self.encoder.encode(screenshot)
    
    def query(self, repr: SceneRepresentation, queries: list[Query]) -> list[Answer]:
        # Parallel query processing
        return [
            self.decoder.answer(repr, q)
            for q in queries  # Parallelizable!
        ]
    
    # Flexible queries enabled:
    # - "Where is the submit button?" -> coordinates
    # - "What text is at (x,y)?" -> OCR
    # - "What elements changed since last frame?" -> diff
    # - "What would clicking here do?" -> action prediction
```

One encode, many queries.

## ROI Calculation
- Screen analysis per action: 500ms current
- Query-based: 50ms (10x faster)
- Actions per session: 100+
- Value: 45s saved per session

## Acceptance Criteria

| AC | Description | Verification |
|----|-------------|--------------|
| AC1 | Single-pass screen encoder produces reusable SceneRepresentation | Unit test: encode once, verify repr is hashable and reusable across queries |
| AC2 | Parallel query decoder handles location/OCR/diff query types | Integration test: batch of 3 query types returns correct results |
| AC3 | Support location queries ("Where is X?") returning coordinates | Test: known UI element, verify coordinates within 5px of ground truth |
| AC4 | Support OCR queries ("What text is at (x,y)?") returning text | Test: known text region, verify exact string match |
| AC5 | Support diff queries ("What changed?") returning element deltas | Test: two screenshots with known diff, verify change set matches |
| AC6 | Response time <100ms for single query after encoding | Benchmark: p95 over 100 queries on encoded screen |

## Proposed Solution

**Architecture**: Two-stage encode-then-query pipeline.

1. **UnifiedScreenEncoder** — Takes a screenshot and produces a dense `SceneRepresentation` (feature grid + element embeddings). Uses a lightweight vision backbone (e.g., MobileViT or distilled ViT-S) fine-tuned on UI screenshots. Encoding happens once per screen state change, amortized across all subsequent queries.

2. **LightweightQueryDecoder** — Takes a `SceneRepresentation` + typed `Query` and produces an `Answer`. Query types:
   - `LocationQuery("submit button")` -> `(x, y, w, h)` bounding box
   - `OCRQuery(x, y)` -> `str` text content
   - `DiffQuery(prev_repr, curr_repr)` -> `list[ElementDelta]` changes
   - `ActionQuery(x, y)` -> `ActionPrediction` (what clicking here would do)
   
   Each query decoder head is a small MLP (<1M params) sharing the same scene representation.

3. **Caching Layer** — `SceneRepresentation` is cached and invalidated only on screen change detection (pixel diff threshold). Consecutive queries on static screens skip encoding entirely.

**Key design decisions**:
- Encode-once architecture: O(1) encoding + O(n) parallel queries vs current O(n) independent analyses
- Query type extensibility via `Query` enum — new query types require only a new decoder head
- Integration with existing axterminator accessibility tree: queries can reference AX elements by role/label

**Implementation order**: SceneRepresentation format → Encoder (ViT backbone) → Location decoder → OCR decoder → Diff decoder → Caching layer → Action decoder

## Dependencies

- Vision backbone model selection (MobileViT or ViT-S, need to benchmark latency vs accuracy)
- UI screenshot training data (can bootstrap from axterminator's existing screenshot capture)
- PyTorch or ONNX Runtime for inference (encoder + decoder heads)
- Existing axterminator screenshot capture and accessibility tree APIs (no changes needed)

## DoD Pre-verification

| DoD Gate | Status | Notes |
|----------|--------|-------|
| § 2 Code Quality | Ready | Python, ruff + mypy enforced |
| § 3 Static Analysis | Ready | `ruff check . && mypy .` |
| § 4 Testing | Ready | Unit tests per decoder head, integration test with real screenshots |
| § 5 Change Safety | Ready | New module, existing axterminator APIs unchanged |
| § 8 Security | N/A | Local-only screen capture, no network, no user data exfiltration |
| § 10 Performance | Ready | Benchmark: <100ms query latency (AC6), <200ms encode latency |
| H1 Search First | Ready | No existing query-based screen engine in codebase |
